airflow/models/baseoperator.py

```diff
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
                 k: rt("{}[{}]".format(attr, k), v, context)
                 for k, v in list(content.items())}
         else:
-            result = content
+            result = self._render_nested_template_fields(content, context)
         return result
 
+    def _render_nested_template_fields(self, content, context):
```

Comment:
Since _rendered_template_object_ids is only relevant to this function can we instead change the sig to:

```suggestion
    def _render_nested_template_fields(self, content, context, seen_oids):
```

Reply:
@ashb : I have some concern about this change; I thought about it, but it implies changing `BaseOperator.render_template` public method signature to propagate theses already `seen_oids` to nested template fields:
``` python
    def render_template(self, attr, content, context):
```
would become:
```python
    def render_template(self, attr, content, context, seen_oids):
```
This is why I preferred storing this set as an instance variable.
If you think this is not an issue, just let me know and I'll do it.

Reply:
I think this diff should work:


```diff
diff --git a/airflow/models/baseoperator.py b/airflow/models/baseoperator.py
index b71f3c4ca1..3e081a5e3d 100644
--- a/airflow/models/baseoperator.py
+++ b/airflow/models/baseoperator.py
@@ -632,7 +633,10 @@ def render_template_from_field(self, attr, content, context, jinja_env):
         Renders a template from a field. If the field is a string, it will
         simply render the string and return the result. If it is a collection or
         nested set of collections, it will traverse the structure and render
-        all elements in it. If the field has another type, it will return it as it is.
+        all elements in it. For any other type, it will recursively render attributes
+        listed in its 'template_fields' attribute (class or instance level attribute)
+        when this 'template_fields' is defined only.
+        Finally returns the rendered field.
         """
         rt = self.render_template
         if isinstance(content, six.string_types):
@@ -644,9 +648,23 @@ def render_template_from_field(self, attr, content, context, jinja_env):
                 k: rt("{}[{}]".format(attr, k), v, context)
                 for k, v in list(content.items())}
         else:
-            result = content
+            result = self._render_nested_template_fields(content, context, seen_oids=set())
         return result
 
+    def _render_nested_template_fields(self, content, context, seen_oids):
+        if id(content) not in seen_oids:
+            seen_oids.add(id(content))
+            try:
+                nested_template_fields = content.template_fields
+            except AttributeError:
+                # content has no inner template fields
+                return content
+
+            for field in nested_template_fields:
+                rendered = self.render_template(field, getattr(content, field), context, seen_oids=seen_oids)
+                setattr(content, field, rendered)
+        return content
+
     def render_template(self, attr, content, context):
         """
         Renders a template either from a file or directly in a field, and returns
```

But I may have missed something.

Reply:
Hi @ashb,

I might be wrong, but I think there is one change missing in your diff: `render_template` signature has to be changed:

```diff
            for field in nested_template_fields:
                rendered = self.render_template(field, getattr(content, field), context, seen_oids=seen_oids)
                setattr(content, field, rendered)
        return content

-     def render_template(self, attr, content, context):
+     def render_template(self, attr, content, context, seen_oids):
         """
         Renders a template either from a file or directly in a field, and returns
```

And this is precisily the change I'm not sure it 's appropriate. What do you think ?

Reply:
Oh yeah, I did miss something.
```diff
            for field in nested_template_fields:
                rendered = self.render_template(field, getattr(content, field), context, seen_oids=seen_oids)
                setattr(content, field, rendered)
        return content

-     def render_template(self, attr, content, context):
+     def render_template(self, attr, content, context, seen_oids=None):
         """
         Renders a template either from a file or directly in a field, and returns
```

And then something like this.

```diff
                 k: rt("{}[{}]".format(attr, k), v, context)
                 for k, v in list(content.items())}
         else:
-            result = content
+            if seen_oids is None:
+                seen_oids = set()
+            result = self._render_nested_template_fields(content, context, seen_oids=seen_oids)
         return result
```

Reply:
Thank you for your feedback. 
I understand you would rather change `BaseOperator.render_template` signature. I will rework this as soon as I can find some time for this (hopefully soon enough!).